isis/tests/MockCsmPlugin.cpp

```diff
@@ -0,0 +1,261 @@
+QMap<std::string, csm::Model*> MockCsmPlugin::m_registeredModels = QMap<std::string, csm::Model*>();
+
+/**
+ * Default constructor
+ */
+MockCsmPlugin::MockCsmPlugin() {
+  // deliberately blank for testing
+}
+
+
+/**
+ * Default destructor
+ */
+MockCsmPlugin::~MockCsmPlugin() {}
+
+
+/**
+ * Gets the name of the plugin.
+ *
+ * @return std::string name of the plugin
+ */
+std::string MockCsmPlugin::getPluginName() const {
+  return MockCsmPlugin::PLUGIN_NAME;
+}
+
+
+/**
+ * Gets the name of the manufacturer of the plugin.
+ *
+ * @return std::string the name of the manufacturer of the plugin
+ */
+std::string MockCsmPlugin::getManufacturer() const {
+  return MockCsmPlugin::MANUFACTURER_NAME;
+}
+
+
+/**
+ * Gets the release date of the plugin.
+ *
+ * @return std::string release date
+ */
+std::string MockCsmPlugin::getReleaseDate() const {
+  return MockCsmPlugin::RELEASE_DATE;
+}
+
+
+/**
+ * Returns the version of CSM the Plugin uses.
+ *
+ * @return csm::Version CSM version plugin uses
+ */
+csm::Version MockCsmPlugin::getCsmVersion() const {
+  return csm::Version(3,0,3);
+}
+
+
+/**
+ * Returns the number of sensor models in the plugin
+ *
+ * @return size_t Number of sensor models in the plugin
+ */
+size_t MockCsmPlugin::getNumModels() const {
+  // This will change as models are loaded into the plugin,
+  // and we cannot access the model registry by index so just say it's empty
+  return 0;
+}
+
+
+/**
+ * Returns the model name at the given index.
+ *
+ * @param modelIndex The index number for the sensor model
+ *
+ * @return std::string model name
+ */
+std::string MockCsmPlugin::getModelName(size_t modelIndex) const {
+  // We cannot access the model registry by index so just return a dummy value
+  return "Dummy Model Name";
+}
+
+
+/**
+ * Returns the sensor model family at the given index.
+ *
+ * @param modelIndex the index number for the sensor model family
+ *
+ * @return std::string sensor model family
+ */
+std::string MockCsmPlugin::getModelFamily(size_t modelIndex) const {
+  // We cannot access the model registry by index so just return a dummy value
+  return "TestModelFamily";
+}
+
+
+/**
+ * Returns the CSM sensor model version for a given model.
+ *
+ * @param modelName the model name
+ *
+ * @return csm::Version the version of the csm sensor model
+ */
+csm::Version MockCsmPlugin::getModelVersion(const std::string& modelName) const {
+  return csm::Version(1,0,0);
+}
+
+
+/**
+ * Tests if the sensor model can be created from a given state.
+ *
+ * @param modelName the model name
+ * @param modelState the model state
+ * @param warnings the warning list
+ *
+ * @return bool if the model can be constructed from the state
+ */
+bool MockCsmPlugin::canModelBeConstructedFromState(const std::string& modelName,
+                                    const std::string& modelState,
+                                    csm::WarningList* warnings) const {
+  return m_registeredModels.contains(modelState);
```

Comment:
I'm a little confused about this. `contains(modelName)` would make more sense to me here.

Reply:
The modelState is the modelName in this case. We want to check the modelState, though because that is what actually contains the data.

Reply:
Could you add a comment that the model state is just the model name here?

Reply:
Isn't `m_registeredPlugin` just a map of model name, to actual model object, though? So how would contains work with a whole modelState here (if it weren't the model name)? I might just be missing the point, though, because it is the model name here, and as such, this is a good test to see if the model can be constructed from the state.